### ISSUE TITLE
(sd-webui-comfyui) Stable api for checkpoint/vae/text encoder custom implementation

### ISSUE THREAD

**Issue by ljleb** [comfyanonymous/ComfyUI#2407]
Hi, I am one of the main maintainer of the [sd-webui-comfyui](https://github.com/ModelSurge/sd-webui-comfyui) extension for a1111.

The extension provides a comfyui node that reuses the webui checkpoint already loaded in memory to allow people to reuse resources:
https://github.com/ModelSurge/sd-webui-comfyui/wiki/Webui-Nodes#webui-checkpoint

The way it has been implemented is very fragile and breaks whenever the implementation of the checkpoint/vae/text encoder (and nesting `ModelPatcher`) classes change. It would be very convenient if it was possible for external code to be written against a more stable interface to provide custom checkpoint sources.

In the case of the extension, we need to be able to copy the latents back and forth between the webui and comfyui processes. This is done by moving the latents to cpu, then serializing them, then copying the serialized tensor over to the other process, then deserializing, then moving to the right device for inference.

This is how the above is implemented currently:
https://github.com/ModelSurge/sd-webui-comfyui/blob/main/lib_comfyui/webui/proxies.py

As you can see, it is a very tedious process and it is very fragile:
https://github.com/ModelSurge/sd-webui-comfyui/issues/195

TL;DR: We need a more stable interface for inference using custom code in checkpoint/vae/text encoder classes. Ideally model patching should also have a way to customize the behavior of model patches to i.e. apply LoRA weights to the webui checkpoint.

**Comment by comfyanonymous** [comment]
You could create an empty comfyui model and set the weights to point to the one from the a1111 model.

**Comment by ljleb** [comment]
Can this work if the weights are in a different process? IIUC this does not really work. Either the latents or the a1111 model have to be copied to the comfyui process for this to work. The former results in unstable code and the latter is very slow.

**Comment by comfyanonymous** [comment]
It's possible to share memory between processes but I don't know how it would work with pytorch.

**Comment by ljleb** [comment]
A while ago I tried many things using the `torch.multiprocessing` module but couldn't get it to work after a long series of trial and error. If it is possible, it must be a very very undocumented and non-discussed niche feature. When I was looking for a checkpoint sharing solution, I looked online for what other people had tried, asked GPT4 for ideas, tried many alternatives, etc. to no avail.

But with the latest versions of sd-webui-comfyui, it's now not really possible to share memory even using `torch.multiprocessing`, since the comfyui process is created using `subprocess` instead of `multiprocessing`. The reason we switched to subprocess is that some custom comfyui nodes people use have requirements that are incompatible with the webui venv, so we have to be able to start comfyui using its own venv, and for that we need a completely different process, which cannot share memory during the child process creation as far as I know.

For this, we had to implement a custom memory sharing mechanism, which relies on `multiprocessing.shared_memory`, or a temporary file, depending on user's preference.

Next year, I can try to find a way to modify the classes to make copying the latents easier to implement in the extension and make it more stable, if you're not completely against the idea.

**Comment by ljleb** [comment]
To be clear, by "next year", I meant this week.

**Comment by Ainaemaet** [comment]
I really hope this is taken more seriously. There are so many things auto1111 can do easily that ComfyUI can't, and vice versa - so having the two work together is imperative imo.
